torch/csrc/jit/tensorexpr/mem_dependency_checker.cpp

```diff
@@ -0,0 +1,1261 @@
+#include <torch/csrc/jit/tensorexpr/mem_dependency_checker.h>
+#include <fstream>
+
+namespace torch {
+namespace jit {
+namespace tensorexpr {
+namespace mem_dependency {
+
+const char* AccessToString(AccessType a) {
+  switch (a) {
+    case AccessType::Input:
+      return "Input";
+    case AccessType::Output:
+      return "Output";
+    case AccessType::Load:
+      return "Load";
+    case AccessType::Store:
+      return "Store";
+    case AccessType::Call:
+      return "Call";
+    case AccessType::AtomicAdd:
+      return "AtomicAdd";
+    default:
+      break;
+  }
+  return "Unknown";
+}
+
+void getDependencyChain(
+    const std::shared_ptr<AccessInfo>& info,
+    DependencySet& dependencies) {
+  if (!dependencies.insert(info).second) {
+    return;
+  }
+
+  for (auto& dep : info->dependencies()) {
+    getDependencyChain(dep.second, dependencies);
+  }
+}
+
+void getDependentsChain(
+    const std::shared_ptr<AccessInfo>& info,
+    DependencySet& dependents) {
+  if (!dependents.insert(info).second) {
+    return;
+  }
+
+  for (auto& dep : info->dependents()) {
+    getDependencyChain(dep.second, dependents);
+  }
+}
+
+// AccessInfo
+
+std::vector<const Expr*> AccessInfo::getIndices() const {
+  std::vector<const Expr*> indices;
+
+  if (expr_) {
+    if (const Load* load = dynamic_cast<const Load*>(expr_)) {
+      indices = load->indices();
+    } else if (
+        const FunctionCall* call = dynamic_cast<const FunctionCall*>(expr_)) {
```

Comment:
If you use `auto` here it'll fit on one line I think and it's ok since the right hand side is a `dynamic_cast`.